apollo-router/src/plugins/demand_control/mod.rs

```diff
@@ -100,9 +150,324 @@ impl Plugin for DemandControl {
         if !self.config.enabled {
             service
         } else {
-            ServiceBuilder::new().service(service).boxed()
+            let strategy = self.strategy_factory.create();
+            ServiceBuilder::new()
+                .checkpoint(move |req: execution::Request| {
+                    req.context.extensions().lock().insert(strategy.clone());
+                    // On the request path we need to check for estimates, checkpoint is used to do this, short-circuiting the request if it's too expensive.
+                    Ok(match strategy.on_execution_request(&req) {
+                        Ok(_) => ControlFlow::Continue(req),
+                        Err(err) => ControlFlow::Break(
+                            execution::Response::builder()
+                                .errors(
+                                    err.into_graphql_errors()
+                                        .expect("must be able to convert to graphql error"),
+                                )
+                                .context(req.context.clone())
+                                .build()
+                                .expect("Must be able to build response"),
+                        ),
+                    })
+                })
+                .map_response(|mut resp: execution::Response| {
+                    let req = resp
+                        .context
+                        .unsupported_executable_document()
+                        .expect("must have document");
+                    let strategy = resp
+                        .context
+                        .extensions()
+                        .lock()
+                        .get::<Strategy>()
+                        .expect("must have strategy")
+                        .clone();
+                    resp.response = resp.response.map(move |resp| {
+                        // Here we are going to abort the stream if the cost is too high
+                        // First we map based on cost, then we use take while to abort the stream if an error is emmitted.
+                        // When we terminate the stream we still want to emit a graphql error, so the error response is emitted first before a termination error.
+                        resp.flat_map(move |resp| {
+                            match strategy.on_execution_response(req.as_ref(), &resp) {
+                                Ok(_) => stream::iter(vec![Ok(resp)]),
```

Comment:
the `flat_map` then iter of a single element vec feels weird, wasn't there another solution? https://docs.rs/futures/latest/futures/stream/trait.StreamExt.html#method.then maybe?

Reply:
I don't think this is possible without using the `flat_map`. We have to generate an extra element in the stream for us to be able to use `take_while` for termination.
Without the extra element we can't emit an error to the caller as `take_while` will terminate the stream immediately.

We could write our own layer, but would prefer to see if there is a perf issue with the existing constructs before doing that.

I tried using some other constructs for the single vec e.g. std::iter::once., but the issue would be that we would have to box to obtain the same typeing on the match arms.

Reply:
[edeade8](https://github.com/apollographql/router/pull/4957/commits/edeade88f742c676829e99ae977c73591a835f67)